src/spatialexperiment/SpatialExperiment.py

```diff
@@ -0,0 +1,814 @@
+__license__ = "MIT"
+
+
+class SpatialExperiment(SingleCellExperiment):
+    """Container class for storing data from spatial -omics experiments, extending
+    :py:class:`~singlecellexperiment.SingleCellExperiment` to provide slots for
+    image data and spatial coordinates.
+
+    In contrast to R, :py:class:`~numpy.ndarray` or scipy matrices are unnamed and do
+    not contain rownames and colnames. Hence, these matrices cannot be directly used as
+    values in assays or alternative experiments. We strictly enforce type checks in these cases.
+    """
+
+    def __init__(
+        self,
+        assays: Dict[str, Any] = None,
+        row_ranges: Optional[GRangesOrGRangesList] = None,
+        row_data: Optional[biocframe.BiocFrame] = None,
+        column_data: Optional[biocframe.BiocFrame] = None,
+        row_names: Optional[List[str]] = None,
+        column_names: Optional[List[str]] = None,
+        metadata: Optional[dict] = None,
+        reduced_dims: Optional[Dict[str, Any]] = None,
+        main_experiment_name: Optional[str] = None,
+        alternative_experiments: Optional[Dict[str, Any]] = None,
+        row_pairs: Optional[Any] = None,
+        column_pairs: Optional[Any] = None,
+        # ============== SpatialExperiment arguments ===============
+        spatial_coords: Optional[biocframe.BiocFrame] = None,
+        img_data: Optional[biocframe.BiocFrame] = None,
+        validate: bool = True,
+    ) -> None:
+        """Initialize a spatial experiment.
+
+        Args:
+            assays:
+                A dictionary containing matrices, with assay names as keys
+                and 2-dimensional matrices represented as either
+                :py:class:`~numpy.ndarray` or :py:class:`~scipy.sparse.spmatrix`.
+
+                Alternatively, you may use any 2-dimensional matrix that has
+                the ``shape`` property and implements the slice operation
+                using the ``__getitem__`` dunder method.
+
+                All matrices in assays must be 2-dimensional and have the
+                same shape (number of rows, number of columns).
+
+            row_ranges:
+                Genomic features, must be the same length as the number of rows of
+                the matrices in assays.
+
+            row_data:
+                Features, must be the same length as the number of rows of
+                the matrices in assays.
+
+                Feature information is coerced to a
+                :py:class:`~biocframe.BiocFrame.BiocFrame`. Defaults to None.
+
+            column_data:
+                Sample data, must be the same length as the number of
+                columns of the matrices in assays. For instances of the
+                ``SpatialExperiment`` class, the sample data must include
+                a column named `sample_id`.
+
+                Sample information is coerced to a
+                :py:class:`~biocframe.BiocFrame.BiocFrame`. Defaults to None.
```

Comment:
I would document that `column_data` must contain a `sample_id` column and must match with an image in `img_data`